### PR TITLE
Add option to protect against csv injection attacks

### PR DESCRIPTION
**Pull request by morcoteg**
Added the option to pass in escape_formulas: true into the #add_row method.

In John Legrand's words, from this issue #624
"""
It would be nice to be able to pass an option to block formula injection. We use this Gem to let clients export tables to excel sheets. This is an unsafe practice because a formula could be injected. There should be an ability to block these injections to OWASP standards. (prepending "'" to anything that starts with something possibly malicious. https://www.owasp.org/index.php/CSV_Injection
"""

### CODE REVIEW

**Comment by noniq** [comment]
Thanks again for the PR!

Two thoughts, looking for the opinions of others:

 * Should the escaping be done in [`Cell#clean_value`](https://github.com/caxlsx/caxlsx/blob/bf2cb814f6/lib/axlsx/workbook/worksheet/cell.rb#L373-L379)? This is where XML escaping is currently happening, and I’m wondering if escaping formulas is conceptually similar.
 * Is `#add_row` the right place to add this option? My gut feeling is that this rather belongs to individual cells, but not sure how an easy-to-use API could look like (having to set this on each cell after using `#add_row` would certainly not be a friendly API, for example).

**Comment by morcoteg** [comment]
Good point! I could look around and see if another place is more appropriate.
Currently at my work, there's a lot of places that build the row array, and then call `#add_row`, so I know we aren't using the API now on a per cell basis. But maybe that's not the intended way to use the gem (our Excel generation was built long long ago). Open to suggestions!

**Comment by noniq** [comment]
Using `#add_row` is in fact the recommended way to add cells :-) 

Maybe it should work similar to the way styles and and types are specified, that is:

```ruby
# Escapes formulas in all cells
ws.add_row ['foo', 'bar', 'baz'], escape_formulas: true

# Escapes formulas in the first cell only
ws.add_row ['foo', 'bar', 'baz'], escape_formulas: [true, false, false]
```

What do you think?

**Comment by morcoteg** [comment]
That'd be good. The security concern was around any cell, so when we patched the method, since any cell could be dynamic user input (in our use case), it was nice to have just `escape_formulas: true` since we didn't have to specify or know what cells could be malicious. But it could be nice to have that kind of control also (outside of just our use case at work), so not super sure, open to suggestions on that. It could be nice since we wouldn't be doing unnecessary formula checking on every cell

**Comment by OpenCoderX** [comment]
Allowing escape_formulas to accept an array also has the benefit of matching the syntax used to specify types.

**Comment by noniq** [comment]
Let’s go with this API then: `escape_formulas: true` works on the whole row, if an array is passed instead it works on each individual cell. Similar to how the `styles: …` and `type: …` options work.

Additionally  I think it should be possible to get and set `escape_formulas` directly on a specific cell. Similar for example to `Cell#type` (and `Cell#type=`). This also means that the value should not be escaped when it is written, but rather when the XML is created. I suppose this could be as simple as making `Cell#is_formula?` aware of this new flag.

**Comment by morcoteg** [comment]
Will do

**Comment by morcoteg** [comment]
@noniq Why can't we set the value when setting `escape_formulas=`? And why do we want to set it only when the XML is being generated? Honest question, as I get more familiar with the `cell_serializer`

**Comment by noniq** [comment]
My idea was to add to `Cell#is_formula?` something like `return false if @escape_formulas`. But not sure if this would already be enough, or if it would still be necessary to additionally prepend a single quote to the formula. Have you tried?

**Comment by morcoteg** [comment]
Let me know if I understand this right - So we should have `Cell#is_formula?` return false if we are escaping formulas when we generate XML. According to this code in `CellSerializer` where the xml is generated, if we return false for `Cell#is_formula?`, we will call `CellSerializer#inline_string_serialization`. Are you recommending doing the escaping also in `CellSerializer#inline_string_serialization`? 
```
def string(cell, str='')
        if cell.is_array_formula?
          array_formula_serialization cell, str
        elsif cell.is_formula?
          formula_serialization cell, str
        elsif !cell.ssti.nil?
          value_serialization 's', cell.ssti, str
        else
          inline_string_serialization cell, str
        end
      end

#####

      def inline_string_serialization(cell, str = '')
        str << 't="inlineStr"><is>'
        run_xml_string cell, str
        str << '</is>'
      end
```

I could do that, if that's what you mean. I could also escape the value when setting `Cell#escape_formulas=` as well. Open to suggestions for the right place to escape in this case when we set `Cell#escape_formulas` :)

**Comment by noniq** [comment]
From my experiments it seems that whether Excel interprets a cell value as formula depends only on the kind of serialization for the cell, not on the actual contents:

Currently caxlsx serializes a cell with value `=A1+A2` as `<f>A2+A3</f>` (because `#is_formula?` returns true for this value). This gets interpreted by Excel as formula

If caxlsx instead handles this value like any other string and serializes it as `<is><t>=A2+A3</t></is>`, then Excel seems to display it as literal string `=A2+A3`.

So apparently it’s not even necessary to prepend `'` to the actual cell value, but simply to adapt the serialization process to explicitly *not* serialize a value as formula if escape flag is set, even if the value starts with `=`.

**Comment by morcoteg** [comment]
Great find. Will update this PR to allow that to happen then, when the XML is generated then :)

File: lib/axlsx/workbook/worksheet/cell.rb [hunk 1]
```diff
@@ -30,6 +30,7 @@ class Cell
     # @option options [String] color an 8 letter rgb specification
     # @option options [Number] formula_value The value to cache for a formula cell.
     # @option options [Symbol] scheme must be one of :none, major, :minor
+    # @option options [Boolean] escape_formulas - Whether or not to escape injected formulas (OWASP standard https://www.owasp.org/index.php/CSV_Injection)
```
Comment (noniq) [inline lib/axlsx/workbook/worksheet/cell.rb]: The reference to CSV injection is not 100% adequate, because we’re not generating CSV data. But it serves as a good in-depth explanation. Maybe reword the comment to something like 

> Whether to treat a value starting with an equal sign as formula (default) or as simple string. Allowing user generated data to be interpreted as formulas can be dangerous (see https://www.owasp.org/index.php/CSV_Injection for details).

File: lib/axlsx/workbook/worksheet/cell.rb [hunk 2]
```diff
@@ -102,6 +105,14 @@ def type=(v)
       self.value = @value unless !defined?(@value) || @value.nil?
     end
 
+    # Whether the cell should escape formulas or not.
```
Comment (noniq) [inline lib/axlsx/workbook/worksheet/cell.rb]: Should be reworded similar to the previous notice, eg. “Whether to treat values …” or something in this direction.

File: lib/axlsx/workbook/worksheet/cell.rb [hunk 3]
```diff
@@ -102,6 +105,14 @@ def type=(v)
...
+    # @return [Boolean]
+    attr_reader :escape_formulas
+
+    def escape_formulas=(v)
+      @escape_formulas = v
```
Comment (noniq) [inline lib/axlsx/workbook/worksheet/cell.rb]: Should we validate that v is a Boolean? Not sure if this is really useful, but on the other hand  many (most?) similar methods seem to have such validations.

File: lib/axlsx/workbook/worksheet/row.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@ module Axlsx
   class Row < SimpleTypedList
     include SerializedAttributes
     include Accessors
-    
+
```
Comment (noniq) [inline lib/axlsx/workbook/worksheet/row.rb]: Please do not mix whitespace-only changes with actual code changes. (You’re welcome to create a distinct commit fixing trailing whitespaces in these files, of course!)

File: lib/axlsx/workbook/worksheet/worksheet.rb [hunk 5]
```diff
@@ -397,6 +397,7 @@ def drawing
     # @option options [Array, Integer] style
     # @option options [Array] widths each member of the widths array will affect how auto_fit behavies.
     # @option options [Float] height the row's height (in points)
+    # @option options [Boolean] escape_formulas - whether or not to escape injected CSV formulas
```
Comment (noniq) [inline lib/axlsx/workbook/worksheet/worksheet.rb]: Type signature should be `[Array, Boolean]`

Please also add an usage example to this comment (simply take one of the existing examples as template).

File: test/workbook/worksheet/tc_cell.rb [hunk 6]
```diff
@@ -321,6 +322,33 @@ def test_to_xml_string_formula
...
+      ], escape_formulas: [true, false, true]
+    end
+    doc = Nokogiri::XML(ws.to_xml_string)
+    doc.remove_namespaces!
+
+    assert(doc.xpath("//t[text()='=IF(2+2=4,4,5)']").any?)
+    assert(doc.xpath("//f[text()='IF(13+13=4,4,5)']").any?)
+    assert(doc.xpath("//t[text()='=IF(99+99=4,4,5)']").any?)
+  end
+
```
Comment (noniq) [inline test/workbook/worksheet/tc_cell.rb]: Perfect test cases! 🥳

File: lib/axlsx/workbook/worksheet/cell.rb (hunk 3 above)
Comment (morcoteg) [reply lib/axlsx/workbook/worksheet/cell.rb]: Yeah, might as well add that constraint, it will help make the API more clear, good call

**Comment by noniq** [comment]
Looks excellent, thank you very much!

While waiting for Travis to finish, would you mind adding an example to https://github.com/caxlsx/caxlsx/blob/master/examples/example.rb ?

I’m actually not sure how this file is meant to be used. It seems to generate example files for various features, and by commenting out some rows where it says `examples << …` one seems to be able to restrict the number of generated files. I suppose the generated files would be useful for manual testing with real applications (Excel, Open Office, …).  

Most of the examples are generated as worksheets in a single workbook (file), but some are generated as separate workbooks (files) – no idea what this distinction is based on. Just do what makes most sense to you. In the end it should be possible to run `example.rb` in way that it produces (maybe amongst others) a file containing some escaped and some unescaped formulas, ready for manual testing.



In the end it should be possible to make `examples.rb` generate a worksheet or workbook that can be opened in

**Comment by morcoteg** [comment]
No problem! Will get on that :) thanks for the great code review comments!

**Comment by noniq** [comment]
Hm … No, let’s do it in this PR, it can be seen as part of the feature’s documentation.

**Comment by morcoteg** [comment]
@noniq added the example, and verified locally that the formulas are being escaped as expected when using `examples/example.rb`

**Comment by noniq** [comment]
Marvellous! Thank you for this valuable contribution, it was a pleasure to cooperate with you! 👏

**Comment by morcoteg** [comment]
Thank you! You as well, great comments that helped a lot!

**Comment by morcoteg** [comment]
Forgot to ask, can I be added as contributor if I add a commit?

**Comment by noniq** [comment]
You’re already listed on https://github.com/caxlsx/caxlsx/graphs/contributors (that list is autogenerated based on all the commits) 😃

**Comment by silambarasanmuthuvel** [comment]
@noniq  @morcoteg @kiskoza Is there any way to set `escape_formulas` =>  `true` as "default value" for every row that we are adding to the Sheet for different implementation (i.e. we are generating Export file in different places)

**Comment by noniq** [comment]
Not possible right now, but it would be a worthy addition! PR welcome! 🙂

**Comment by tagliala** [comment]
> Not possible right now, but it would be a worthy addition! PR welcome! 🙂

Hi, would it be a better default choice to have `escape_formulas` set to true and release a new major version because of the breaking change? Are you interested in a PR?

**Comment by kiskoza** [comment]
I like the idea of having `Workbook#escape_formulas` which can set the default for `Cell#escape_formulas`.

For now I would make it default to `:unset` (or similar, I'm bad with naming) with a deprecation warning that says it's working as `false` for now, but in the future it'll be `true` and it may break your code if you're not setting an exlicit value.

I wouldn't release v4.0 just for this change, but we can open up a new discussion and figure out a roadmap to v4.0, so we don't end up needing v5.0 a few weeks later
